src/listmode_buildblock/CListModeDataSAFIR.cxx

```diff
@@ -0,0 +1,134 @@
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+*/
+
+#include <iostream>
+#include <fstream>
+#include <typeinfo>
+
+#include "stir/ExamInfo.h"
+#include "stir/Succeeded.h"
+#include "stir/is_null_ptr.h"
+
+#include "boost/static_assert.hpp"
+#include "boost/pointer_cast.hpp"
+
+#include "stir/listmode/CListModeDataSAFIR.h"
+
+
+//#include "stir/ExamInfo.h"
+
+
+#ifndef STIR_NO_NAMESPACES
+using std::cerr;
+using std::endl;
+using std::ios;
+using std::fstream;
+using std::ifstream;
+using std::istream;
+#endif
+
+START_NAMESPACE_STIR;
+
+
+template <class CListRecordT>
+CListModeDataSAFIR<CListRecordT>::
+CListModeDataSAFIR(const std::string& listmode_filename, const std::string& crystal_map_filename, const std::string& template_proj_data_filename)
+  : listmode_filename(listmode_filename), map(boost::make_shared<ListEventRecordMapFromFile>(crystal_map_filename))
+{
+	this->exam_info_sptr.reset(new ExamInfo);
+
+	// Here we are reading the scanner data from the template projdata
+	shared_ptr<ProjData> template_proj_data_ptr =
+//			ProjData::read_from_file("muppet.hs");
+			ProjData::read_from_file(template_proj_data_filename);
+	scanner_sptr = shared_ptr<Scanner>( new Scanner(*template_proj_data_ptr->get_proj_data_info_ptr()->get_scanner_ptr()));
+
+	if( open_lm_file() == Succeeded::no )
+	{
+		error("CListModeDataSAFIR: Could not open listmode file %s\n", listmode_filename.c_str());
+	}
+}
+
+
+template <class CListRecordT>
+std::string
+CListModeDataSAFIR<CListRecordT>::
+get_name() const
+{
+	return listmode_filename;
+}
+
+template <class CListRecordT>
+shared_ptr <CListRecord>
+CListModeDataSAFIR<CListRecordT>::
+get_empty_record_sptr() const
+{
+	boost::shared_ptr<CListRecordSAFIR> sptr(new CListRecordSAFIR);
+	sptr->event_SAFIR().set_map(map);
+	return boost::static_pointer_cast<CListRecord>(sptr);
+}
+
+template <class CListRecordT>
+Succeeded
+CListModeDataSAFIR<CListRecordT>::
+get_next_record(CListRecord& record_of_general_type) const
+{
+	CListRecordT& record = static_cast<CListRecordT&>(record_of_general_type);
+	Succeeded status = current_lm_data_ptr->get_next_record(record);
+	if( status == Succeeded::yes ) record.event_SAFIR().set_map(map);
+	return status;
+	
+}
+
+template <class CListRecordT>
+Succeeded
+CListModeDataSAFIR<CListRecordT>::
+reset()
+{
+	return current_lm_data_ptr->reset();
+}
+
+template <class CListRecordT>
+Succeeded
+CListModeDataSAFIR<CListRecordT>::
+open_lm_file() const
+{
+	cerr << "CListModeDataSAFIR: opening file " << listmode_filename << endl;
+	shared_ptr<istream> stream_ptr(new fstream(listmode_filename.c_str(), ios::in | ios::binary ));
+	if(!(*stream_ptr))
+	{
+		warning("CListModeDataSAFIR: cannot open file %s\n", listmode_filename.c_str());
+		return Succeeded::no;
+	}
+	stream_ptr->seekg(32);
```

Comment:
As far as I could find out, the C++ standard requires std::streamoff to be a typedef to a signed integral type (system-dependend), that means it's implicitely convertible. I have put a cast operator nevertheless.